### PR TITLE
[Feature] Video optimization step 1

### PR DESCRIPTION
**Pull request by daniloercoli**
This PR adds and uses the [Media for Mobile library](https://github.com/INDExOS/media-for-mobile) in order to compress and resize video files before uploading to the server.

We need to test video re-encoding capabilities on a large set of devices before shipping it to our end-users, so basically the plan to test this new feature/library is the following:
- Force enable video optimization to alpha releases
- Force enable video optimization on any debug build

Our end users will not see this feature until we decide is stable enough, and we've completed the integration of it in Settings screen. I suppose there will be a new setting for it.

Now the process of re-encoding a video is a blocking task, there is a progress dialog in place that is shown until the video is processed. Then, the resulting video is added to the visual editor and the upload starts. 
In our final implementation it's better to integrate video re-encoding in the async branch only. Once we've tested that the library works fine.

Videos are re-encoded to 720p (if they're larger) 30 frame per sec, 3000 KBytes rate. 

Next steps (step 2) is to start working in the settings screen, and provide the ability to turn this feature ON/OFF (and select different bitrate?).

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -1743,6 +1748,150 @@ public boolean addMedia(Uri mediaUri) {
...
+        if (progressDialog != null && progressDialog.isShowing()) {
+            try {
+                progressDialog.dismiss();
+            } catch (IllegalArgumentException e) {
+                // dialog doesn't exist
+            }
+        }
+    }
+
+    private class VideoOptimizationProgressListener implements org.m4m.IProgressListener {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: What do you think about moving all the video optimization code to a separate class? My concern is that `EditPostActivity` is getting very large and unwieldy.

File: WordPress/src/main/java/org/wordpress/android/util/WPVideoUtils.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
+package org.wordpress.android.util;
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/util/WPVideoUtils.java]: This class would benefit from much more commenting. That would make it much easier for another dev to work on it.

File: WordPress/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -1693,4 +1693,10 @@
     <!-- smart toasts -->
     <string name="smart_toast_photo_long_press">Tap and hold to select multiple photos</string>
     <string name="smart_toast_comments_long_press">Tap and hold to select multiple comments</string>
+
+
+    <!-- Video processing -->
+    <string name="video_optimization_in_progress">Optimizing your video&#8230;</string>
+    <string name="video_optimization_cant_optimize">Sorry, we can\'t optimize this video. Using the original file.</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: We don't appear to use `video_optimization_cant_optimize` anywhere.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 4]
```diff
@@ -1743,6 +1748,150 @@ public boolean addMedia(Uri mediaUri) {
...
+        public void onError(Exception exception) {
+            AppLog.e(AppLog.T.MEDIA, "Can't optimize the video", exception);
+            if (isStopped || isFinishing()) {
+                return;
+            }
+            ProgressDialog pd = mWeakProgressDialog.get();
+            dismissProgressDialog(pd);
+            Context context = mWeakContext.get();
+            if (context != null) {
+                ToastUtils.showToast(context, R.string.video_optimization_generic_error_message, Duration.LONG);
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I'm wondering if this toast is necessary. The user can't do anything about it, and I can't think of why they'd need to know this information.

File: WordPress/src/main/java/org/wordpress/android/util/WPVideoUtils.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    private static final int OUTPUT_WIDTH = 1280;
+    private static final int OUTPUT_HEIGHT = 720;
+    private static final int OUTPUT_BITRATE_KB = 3000;
+
+    // Default parameters for the audio encoder
+    private static final String AUDIO_MIME_TYPE = "audio/mp4a-latm";
+    private static final int AUDIO_OUTPUT_BIT_RATE = 96 * 1024;
+
+    public static MediaComposer getVideoOptimizationComposer(Context ctx, String inputFile, String outFile, org.m4m.IProgressListener listener) {
+        if (ctx == null || inputFile == null || listener == null) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/util/WPVideoUtils.java]: We could skip these null checks and simply make the parameters `@NonNull`.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 4 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: It's there just to inform our alpha testers of the error occurred compressing the video, but I agree we can remove it and add a Crashlytics log instead to gather stats about compressions problems with videos.

File: WordPress/src/main/res/values/strings.xml (hunk 3 above)
Comment (daniloercoli) [inline WordPress/src/main/res/values/strings.xml]: Right, it's not used anymore. Forgot to remove it.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Good call, will move this logic to `WPVideoUtils`.

**Comment by daniloercoli** [comment]
@nbradbury - Ready for another pass!

**Comment by nbradbury** [comment]
Looks good! :shipit:

**Pull request merged**